sw/device/lib/usbdev.h

```diff
@@ -224,10 +224,17 @@ void usbdev_endpoint_setup(usbdev_ctx_t *ctx, int ep, int enableout,
  */
 void usbdev_init(usbdev_ctx_t *ctx);
 
-// Used for tracing what is going on
+// Used for tracing what is going on, may impact timing.
+//#define ENABLE_TRC
+#ifdef ENABLE_TRC
 #include "sw/device/lib/uart.h"
 #define TRC_S(s) uart_send_str(s)
 #define TRC_I(i, b) uart_send_uint(i, b)
 #define TRC_C(c) uart_send_char(c)
+#else
+#define TRC_S(s)
+#define TRC_I(i, b)
+#define TRC_C(c)
+#endif
```

Comment:
I noted that if any of this tracing is done, the callback functions in the hello_usbdev example are never called and we don't see any characters on UART received over USB.

Reply:
That's probably a latency issue. The usbdpi module is quite fragile: If the device takes too long with certain responses, the transaction will not complete correctly. This might be related to the addition of the SPI to the usbdev_poll loop.

This should not cause issues with a real USB host, as the latencies are quite relaxed (>=50 ms) in that case.